File: windows/Gantry.Windows/Services/ElegooCc1Client.cs
using System.IO;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using Gantry.Models;

namespace Gantry.Services;

public sealed class ElegooCc1Client : IPrinterConnection
{
    private readonly SavedPrinter _printer; private readonly Action<MqttEvent> _onEvent;
    private readonly CancellationTokenSource _cts = new(); private readonly SemaphoreSlim _sendGate = new(1, 1); private ClientWebSocket? _socket;
    private PrinterTelemetry _telemetry = new();
    /// <summary>Cmd 386 handshake and the single stream slot, shared by every camera viewer of this printer.</summary>
    public ElegooVideoGate VideoGate { get; }
    public ElegooCc1Client(SavedPrinter printer, Action<MqttEvent> onEvent)
    {
        _printer = printer; _onEvent = onEvent;
        VideoGate = new ElegooVideoGate((enable, requestId) => SendMethod(386, new { Enable = enable ? 1 : 0 }, requestId));
    }
    public void Start() => _ = Task.Run(RunAsync);
    public void Stop() { _cts.Cancel(); _socket?.Dispose(); }

    public void SendMethod(int method, object? parameters = null, string? requestId = null) => _ = Task.Run(async () =>
    {
        if (_socket?.State != WebSocketState.Open) return;
        var message = new { Id = _printer.Serial, Data = new {
            Cmd = method, Data = parameters ?? new { }, RequestID = requestId ?? Guid.NewGuid().ToString("N"),
            MainboardID = _printer.Serial, TimeStamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(), From = 1 },
            Topic = $"sdcp/request/{_printer.Serial}" };
        var bytes = JsonSerializer.SerializeToUtf8Bytes(message);
        try
        {
            await _sendGate.WaitAsync(_cts.Token);
            try { await _socket.SendAsync(bytes, WebSocketMessageType.Text, true, _cts.Token); }
            finally { _sendGate.Release(); }
        }
        catch { }
    });

    private async Task RunAsync()
    {
        try
        {
            _socket = new ClientWebSocket();
            await _socket.ConnectAsync(new Uri($"ws://{_printer.Host}:{_printer.Port ?? 3030}/websocket"), _cts.Token);
            VideoGate.Reset();
            _onEvent(new MqttEvent { Type = MqttEventType.Connected }); SendMethod(0); SendMethod(1); SendMethod(512, new { TimePeriod = 5000 }); _ = PollStatusAsync();
            var chunk = new byte[65536]; using var message = new MemoryStream();
            while (!_cts.IsCancellationRequested)
            {
                var result = await _socket.ReceiveAsync(chunk, _cts.Token);
                if (result.MessageType == WebSocketMessageType.Close) break;
                message.Write(chunk, 0, result.Count);
                if (!result.EndOfMessage) continue;
                var frame = message.ToArray(); message.SetLength(0);
                if (ElegooVideoGate.ParseReply(frame) is { } reply) VideoGate.HandleReply(reply.RequestId, reply.Ack);
                var updated = ElegooStatusParser.Cc1(frame, _telemetry);
                if (updated is not null) { _telemetry = updated; _onEvent(new MqttEvent { Type = MqttEventType.Telemetry, Telemetry = updated }); }
            }
            VideoGate.Reset();
            if (!_cts.IsCancellationRequested) _onEvent(new MqttEvent { Type = MqttEventType.Disconnected });
        }
        catch (OperationCanceledException) { }
        catch (Exception error) { VideoGate.Reset(); if (!_cts.IsCancellationRequested) _onEvent(new MqttEvent { Type = MqttEventType.Disconnected, Reason = error.Message }); }
    }

    private async Task PollStatusAsync()
    {
        try { while (!_cts.IsCancellationRequested) { await Task.Delay(TimeSpan.FromSeconds(10), _cts.Token); SendMethod(0); } }
        catch (OperationCanceledException) { }
    }
}

File: Sources/Gantry/Services/ElegooCC1Client.swift
import Foundation

final class ElegooCC1Client: PrinterConnection, @unchecked Sendable {
    private let printer: SavedPrinter
    private let onEvent: @Sendable (MQTTClient.Event) -> Void
    private let queue: DispatchQueue
    private var task: URLSessionWebSocketTask?
    private var statusTimer: DispatchSourceTimer?
    private var telemetry = PrinterTelemetry()
    private var stopped = false

    init(printer: SavedPrinter, onEvent: @escaping @Sendable (MQTTClient.Event) -> Void) {
        self.printer = printer; self.onEvent = onEvent
        queue = DispatchQueue(label: "pl.gantry.elegoo.cc1.\(printer.serial)")
    }

    func start() { queue.async { [weak self] in self?.connect() } }
    func stop() { queue.async { [weak self] in
        self?.stopped = true; self?.statusTimer?.cancel(); self?.statusTimer = nil
        self?.task?.cancel(with: .goingAway, reason: nil); self?.task = nil
    } }

    func sendMethod(_ command: Int, data: [String: Any] = [:]) {
        guard let parameters = try? JSONSerialization.data(withJSONObject: data) else { return }
        queue.async { [weak self] in
            guard let self, let task else { return }
            let decoded = (try? JSONSerialization.jsonObject(with: parameters)) as? [String: Any] ?? [:]
            let request: [String: Any] = [
                "Id": printer.serial,
                "Data": ["Cmd": command, "Data": decoded, "RequestID": UUID().uuidString.replacingOccurrences(of: "-", with: ""),
                         "MainboardID": printer.serial, "TimeStamp": Int(Date().timeIntervalSince1970 * 1000), "From": 1],
                "Topic": "sdcp/request/\(printer.serial)"
            ]
            guard let json = try? JSONSerialization.data(withJSONObject: request), let text = String(data: json, encoding: .utf8) else { return }
            task.send(.string(text)) { _ in }
        }
    }

    private func connect() {
        stopped = false
        guard let url = URL(string: "ws://\(printer.host):\(printer.port ?? 3030)/websocket") else { return }
        let task = URLSession.shared.webSocketTask(with: url); self.task = task; task.resume()
        onEvent(.connected); sendMethod(0); sendMethod(1); sendMethod(512, data: ["TimePeriod": 5000]); startStatusTimer(); receive()
    }

    private func startStatusTimer() {
        statusTimer?.cancel()
        let timer = DispatchSource.makeTimerSource(queue: queue)
        timer.schedule(deadline: .now() + 10, repeating: 10)
        timer.setEventHandler { [weak self] in self?.sendMethod(0) }
        timer.resume(); statusTimer = timer
    }

    private func receive() {
        task?.receive { [weak self] result in
            guard let self else { return }
            self.queue.async {
                switch result {
                case .failure(let error):
                    self.statusTimer?.cancel(); self.statusTimer = nil
                    if !self.stopped { self.onEvent(.disconnected(error.localizedDescription)) }
                case .success(let message):
                    let data: Data
                    switch message { case .data(let value): data = value; case .string(let value): data = Data(value.utf8); @unknown default: data = Data() }
                    if let updated = ElegooStatusParser.cc1(data: data, previous: self.telemetry) {
                        self.telemetry = updated; self.onEvent(.telemetry(updated))
                    }
                    if !self.stopped { self.receive() }
                }
            }
        }
    }
}
